ttnn/cpp/ttnn/tensor/layout/tensor_layout.cpp

```diff
@@ -122,29 +151,82 @@ size_t TensorLayout::compute_page_size_bytes(const Size& page_size) const {
     return page_config_.get_page_size_bytes(page_size, dtype_);
 }
 
+Size TensorLayout::compute_logical_shard_shape(const ttnn::SimpleShape& shape) const {
+    if (shard_shape_.has_value()) {
+        return shard_shape_.value();
+    }
+    const int rank = static_cast<int>(shape.rank());
+
+    size_t shard_width = 1;
+    size_t shard_height = 1;
+
+    // Iterate dims in reverse order
+    // Even tensor of rank 0 or 1
+    for (int i = -1; i >= -2; --i) {
+        if (i >= -rank) {
+            auto& dim = i == -1 ? shard_width : shard_height;
+            dim *= shape[i];
+        }
+    }
+
+    return Size{shard_height, shard_width};
+}
+
+Size TensorLayout::compute_physical_shard_shape(const ttnn::SimpleShape& shape) const {
+    const auto& logical_shard_shape = compute_logical_shard_shape(shape);
+    return Size{round_up(logical_shard_shape.height(), alignment_[-2]), round_up(logical_shard_shape.width(), alignment_[-1])};
+}
+
 Size TensorLayout::compute_physical_shape(const ttnn::SimpleShape& shape) const {
     const int rank = static_cast<int>(shape.rank());
     const int alignment_rank = static_cast<int>(alignment_.size());
-    const int max_rank = std::max(rank, alignment_rank);
+
     size_t width = 1;
     size_t height = 1;
 
-    // Iterate dims in reverse order and ensure alignment
-    // Even tensor of rank 0 or 1 must be aligned (to Tile / Page / Shard)
-    for (int i = -1; i >= -max_rank; --i) {
-        auto& dim = i == -1 ? width : height;
-        if(i >= -rank) {
-            dim *= shape[i];
+    if (alignment_rank > 2) {
```

Comment:
Why do we have a different handling for `alignment_rank > 2` and `alignment_rank <= 2`?

Reply:
Idea is that if alignment_rank < 2, then I can treat it as shard -> alignment. Otherwise, I cannot and must fallback to old alignment logic